src/validate.py

```
@@ -343,69 +344,73 @@ def loop(validation_args_file: str, task_id: str = None, auto_clean_cache: bool
-                    assignment_id=resp["id"],
-                    local_test=False,
-                )
-                break  # Break the loop if no exception
-            except KeyboardInterrupt:
-                break
-            except Exception as e:
-                logger.error(f"Attempt {attempt + 1} failed: {e}")
-                if attempt == 2:
-                    logger.error(
-                        f"Marking assignment {assignment_id} as failed after 3 attempts"
+            resp = resp.json()
+            eval_file = download_file(resp["data"]["validation_set_url"])
+            assignment_id = resp["id"]
+
+            for attempt in range(3):
+                try:
+                    ctx = click.Context(validate)
+                    ctx.invoke(
+                        validate,
+                        model_name_or_path=resp["task_submission"]["data"]["hg_repo_id"],
+                        base_model=resp["data"]["base_model"],
+                        eval_file=eval_file,
+                        context_length=resp["data"]["context_length"],
+                        max_params=resp["data"]["max_params"],
+                        validation_args_file=validation_args_file,
+                        assignment_id=resp["id"],
+                        local_test=False,
                     )
-                    fed_ledger.mark_assignment_as_failed(assignment_id)
-
-        os.remove(eval_file)
-
+                    break  # Break the loop if no exception
+                except KeyboardInterrupt:
+                    break
+                except Exception as e:
+                    logger.error(f"Attempt {attempt + 1} failed: {e}")
+                    if attempt == 2:
+                        logger.error(
+                            f"Marking assignment {assignment_id} as failed after 3 attempts"
+                        )
+                        fed_ledger.mark_assignment_as_failed(assignment_id)
```

If it's caused by that weird device side error, we should not mark this task as failed. So we need to raise that exception from `validate` side as well to handle it.

---

Updated this